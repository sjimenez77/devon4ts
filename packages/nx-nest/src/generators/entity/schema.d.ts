export interface EntityGeneratorSchema {
  name: string;
  projectName: string;
  module?: string;
}
