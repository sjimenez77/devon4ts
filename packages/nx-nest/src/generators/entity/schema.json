{
  "$schema": "https://json-schema.org/schema",
  "$id": "Entity",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "projectName": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 1
      },
      "x-prompt": "In which project would you like to add the entity?"
    },
    "module": {
      "type": "string",
      "description": "Module where the entity will be generated"
    }
  },
  "required": ["name", "projectName"]
}
