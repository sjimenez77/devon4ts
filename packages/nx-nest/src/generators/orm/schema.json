{
  "$schema": "http://json-schema.org/schema",
  "$id": "InitTypeormSchematic",
  "title": "Init Typeorm Schema",
  "type": "object",
  "properties": {
    "orm": {
      "type": "string",
      "enum": ["typeorm"],
      "default": "typeorm"
    },
    "db": {
      "type": "string",
      "enum": ["postgres", "cockroachdb", "mariadb", "mysql", "sqlite", "oracle", "mssql", "mongodb"],
      "default": "sqlite"
    },
    "projectName": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "In which project would you like to add typeorm configuration?"
    }
  },
  "required": ["projectName"]
}
