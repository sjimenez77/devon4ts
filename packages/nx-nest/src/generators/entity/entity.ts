import {
  formatFiles,
  generateFiles,
  installPackagesTask,
  names,
  output,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import * as path from 'path';
import { ASTFileBuilder } from '../../utils/ast-file-builder';
import { findModuleFile } from '../../utils/tree-utils';
import { EntityGeneratorSchema } from './schema';

export async function entityGenerator(tree: Tree, options: EntityGeneratorSchema): Promise<() => void> {
  const appConfig = readProjectConfiguration(tree, options.projectName);
  const projectRoot = `${appConfig.root}/src/${appConfig.projectType === 'application' ? 'app' : 'lib'}`;
  const entityNames = names(options.name);

  const entityPath: string = options.module ?? '.';

  if (options.module && !tree.exists(path.join(projectRoot, entityPath, options.module + '.module.ts'))) {
    throw Error('wrong module');
  }

  generateFiles(tree, path.join(__dirname, 'files'), path.join(projectRoot, entityPath), {
    name: entityNames.fileName,
    className: entityNames.className,
  });
  addEntityToModule(tree, path.join(projectRoot, entityPath), options.name);
  await formatFiles(tree);
  return () => {
    installPackagesTask(tree);
    output.log({
      title: `Entity ${entityNames.className} created.`,
      bodyLines: [
        'Please update your DataSource configuration by adding the entity into entities array',
        'You can find it at CoreModule',
      ],
    });
  };
}

function addEntityToModule(tree: Tree, modulePath: string, entityName: string): void {
  const moduleFile = findModuleFile(tree, modulePath);

  if (!moduleFile) {
    return;
  }

  const contentBuilder = new ASTFileBuilder(tree.read(path.join(modulePath, moduleFile))!.toString());
  const moduleName = contentBuilder.getModuleClassName();
  const entityNames = names(entityName);

  if (!moduleName) {
    return;
  }

  contentBuilder
    .addImports(entityNames.className, './' + entityNames.fileName + '.entity')
    .addImports('TypeOrmModule', '@nestjs/typeorm')
    .addTypeormFeatureToModule(moduleName, entityNames.className);

  tree.write(path.join(modulePath, moduleFile), contentBuilder.build());
}

export default entityGenerator;
