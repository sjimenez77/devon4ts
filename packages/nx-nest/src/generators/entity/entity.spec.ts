import { ProjectConfiguration, Tree, readProjectConfiguration } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { ApplicationGeneratorOptions } from '@nx/nest/src/generators/application/schema';
import libraryGenerator from '@nx/nest/src/generators/library/library';
import { LibraryGeneratorOptions } from '@nx/nest/src/generators/library/schema';
import { moduleGenerator } from '@nx/nest/src/generators/module/module';
import { join } from 'path';
import applicationGenerator from '../application/application';
import { entityGenerator } from './entity';

describe('entity generator', () => {
  describe('application', () => {
    let tree: Tree;
    const options: ApplicationGeneratorOptions = {
      name: 'test',
      projectNameAndRootFormat: 'as-provided',
      directory: 'apps/test',
    };
    let config: ProjectConfiguration;
    beforeEach(async () => {
      tree = createTreeWithEmptyWorkspace();
      await applicationGenerator(tree, options);
      config = readProjectConfiguration(tree, 'test');
    }, 60000);

    it('should generate the entity', async () => {
      expect(config).toBeDefined();
      await entityGenerator(tree, { name: 'testo', projectName: options.name });
      expect(tree.exists(join(config.root, 'src/app/testo.entity.ts'))).toBeTruthy();
      expect(tree.read(join(config.root, 'src/app/testo.entity.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should update the library module if no other module is specified', async () => {
      await entityGenerator(tree, { name: 'testo', projectName: options.name });
      expect(tree.exists(join(config.root, 'src/app/app.module.ts'))).toBeTruthy();
      expect(tree.read(join(config.root, 'src/app/app.module.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should update the specified module', async () => {
      await moduleGenerator(tree, {
        name: 'modulero',
        directory: 'apps/test/src/app/modulero',
        nameAndDirectoryFormat: 'as-provided',
      });
      await entityGenerator(tree, { name: 'testo', projectName: options.name, module: 'modulero' });
      expect(tree.read(join(config.root, 'src/lib/modulero/modulero.module.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should throw an error if the specified module doesnt exists', async () => {
      await expect(
        entityGenerator(tree, { name: 'testo', projectName: options.name, module: 'no-exists' }),
      ).rejects.toThrow(Error);
    });
  });

  describe('library', () => {
    let tree: Tree;
    const options: LibraryGeneratorOptions = {
      name: 'test',
      projectNameAndRootFormat: 'as-provided',
      directory: 'apps/test',
    };
    let config: ProjectConfiguration;
    beforeEach(async () => {
      tree = createTreeWithEmptyWorkspace();
      await libraryGenerator(tree, options);
      config = readProjectConfiguration(tree, 'test');
    }, 60000);

    it('should generate the entity', async () => {
      expect(config).toBeDefined();
      await entityGenerator(tree, { name: 'testo', projectName: options.name });
      expect(tree.exists(join(config.root, 'src/lib/testo.entity.ts'))).toBeTruthy();
      expect(tree.read(join(config.root, 'src/lib/testo.entity.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should update the library module if no other module is specified', async () => {
      await entityGenerator(tree, { name: 'testo', projectName: options.name });
      expect(tree.read(join(config.root, 'src/lib/test.module.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should update the specified module', async () => {
      await moduleGenerator(tree, {
        name: 'modulero',
        directory: 'apps/test/src/lib/modulero',
        nameAndDirectoryFormat: 'as-provided',
      });
      await entityGenerator(tree, { name: 'testo', projectName: options.name, module: 'modulero' });
      expect(tree.read(join(config.root, 'src/lib/modulero/modulero.module.ts'))?.toString('utf-8')).toMatchSnapshot();
    });

    it('should throw an error if the specified module doesnt exists', async () => {
      await expect(
        entityGenerator(tree, { name: 'testo', projectName: options.name, module: 'no-exists' }),
      ).rejects.toThrow(Error);
    });
  });
});
