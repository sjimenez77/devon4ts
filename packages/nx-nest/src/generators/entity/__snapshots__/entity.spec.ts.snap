// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity generator application should generate the entity 1`] = `
"import { CreateDateColumn, Entity, PrimaryGeneratedColumn, UpdateDateColumn, VersionColumn } from 'typeorm';

@Entity()
export class Testo {
  @PrimaryGeneratedColumn('increment')
  id!: number;

  @VersionColumn({ default: 1 })
  version!: number;

  @CreateDateColumn()
  createdAt!: string;

  @UpdateDateColumn()
  updatedAt!: string;
}
"
`;

exports[`entity generator application should update the library module if no other module is specified 1`] = `
"import { Module } from '@nestjs/common';
import { CoreModule } from './core/core.module';
import { Testo } from './testo.entity';
import { TypeOrmModule } from '@nestjs/typeorm';

@Module({
  imports: [TypeOrmModule.forFeature([Testo]), CoreModule],
  controllers: [],
  providers: [],
})
export class AppModule {}
"
`;

exports[`entity generator application should update the specified module 1`] = `undefined`;

exports[`entity generator library should generate the entity 1`] = `
"import {
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  VersionColumn,
} from 'typeorm';

@Entity()
export class Testo {
  @PrimaryGeneratedColumn('increment')
  id!: number;

  @VersionColumn({ default: 1 })
  version!: number;

  @CreateDateColumn()
  createdAt!: string;

  @UpdateDateColumn()
  updatedAt!: string;
}
"
`;

exports[`entity generator library should update the library module if no other module is specified 1`] = `
"import { Module } from '@nestjs/common';
import { Testo } from './testo.entity';
import { TypeOrmModule } from '@nestjs/typeorm';

@Module({
  controllers: [],
  providers: [],
  exports: [],
  imports: [TypeOrmModule.forFeature([Testo])],
})
export class TestModule {}
"
`;

exports[`entity generator library should update the specified module 1`] = `
"import { Module } from '@nestjs/common';
import { Testo } from './testo.entity';
import { TypeOrmModule } from '@nestjs/typeorm';

@Module({
  imports: [TypeOrmModule.forFeature([Testo])],
})
export class ModuleroModule {}
"
`;
